<?xml version="1.0" encoding="utf-8"?>
<!-- =================================================================================== -->
<!-- |    This document was autogenerated by xacro from wamv_gazebo.urdf.xacro         | -->
<!-- |    EDITING THIS FILE BY HAND IS NOT RECOMMENDED                                 | -->
<!-- =================================================================================== -->
<!-- Basic WAM-V with gazebo plugins for dynamics -->
<robot name="WAM-V">
  <link name="wamv/left_battery_link">
    <visual name="left_battery_visual">
      <origin rpy="0 0 0" xyz="0 -1.03 -.45"/>
      <geometry>
        <mesh filename="package://vrx_gazebo/models/battery/mesh/battery.dae"/>
      </geometry>
    </visual>
    <collision name="left_battery_collision">
      <origin rpy="0 0 0" xyz="0 0 .15"/>
      <geometry>
        <box size=".6 .4 .31"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy=" 0 0 0" xyz="0 0 .15"/>
      <mass value="23.5"/>
      <inertia ixx="0.5015291666666667" ixy="0" ixz="0" iyy="0.8931958333333333" iyz="0" izz="1.0183333333333333"/>
    </inertial>
  </link>
  <joint name="wamv/left_chasis_battery_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0 1 0.45"/>
    <parent link="wamv/base_link"/>
    <child link="wamv/left_battery_link"/>
  </joint>
  <link name="wamv/right_battery_link">
    <visual name="right_battery_visual">
      <origin rpy="0 0 0" xyz="0 -1.03 -.45"/>
      <geometry>
        <mesh filename="package://vrx_gazebo/models/battery/mesh/battery.dae"/>
      </geometry>
    </visual>
    <collision name="right_battery_collision">
      <origin rpy="0 0 0" xyz="0 0 .15"/>
      <geometry>
        <box size=".6 .4 .31"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy=" 0 0 0" xyz="0 0 .15"/>
      <mass value="23.5"/>
      <inertia ixx="0.5015291666666667" ixy="0" ixz="0" iyy="0.8931958333333333" iyz="0" izz="1.0183333333333333"/>
    </inertial>
  </link>
  <joint name="wamv/right_chasis_battery_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0 -1 0.45"/>
    <parent link="wamv/base_link"/>
    <child link="wamv/right_battery_link"/>
  </joint>
  <!-- Required arg for melodic and kinetic... -->
  <link name="wamv/base_link">
  </link>
  <joint name="wamv/dummy_joint" type="fixed">
    <parent link="wamv/base_link"/>
    <child link="wamv/dummy_link"/>
  </joint>
  <!-- Basic model of the 16' WAM-V USV -->
  <link name="wamv/dummy_link">
    <visual>
      <geometry>
        <mesh filename="package://wamv_description/models/WAM-V-Base/mesh/WAM-V-Base.dae"/>
      </geometry>
    </visual>
    <!-- Main float -->
    <collision name="left_float">
      <origin rpy="0 1.57 0" xyz="-0.4 1.03 0.2"/>
      <geometry>
        <cylinder length="4" radius="0.2"/>
      </geometry>
    </collision>
    <collision name="left_mid_float">
      <origin rpy="0 1.38 0" xyz="1.85 1.03 0.3"/>
      <geometry>
        <cylinder length="0.5" radius="0.17"/>
      </geometry>
    </collision>
    <collision name="left_front_float">
      <origin rpy="0 1.3 0" xyz="2.3 1.03 0.4"/>
      <geometry>
        <cylinder length="0.45" radius="0.12"/>
      </geometry>
    </collision>
    <!-- Front beam -->
    <collision name="front_left_beam_lower">
      <origin rpy="0.78 0 0" xyz="0.9 0.85 1"/>
      <geometry>
        <cylinder length="0.5" radius="0.04"/>
      </geometry>
    </collision>
    <collision name="front_left_beam_upper">
      <origin rpy="1.35 0 0" xyz="0.9 0.6 1.18"/>
      <geometry>
        <cylinder length="0.2" radius="0.04"/>
      </geometry>
    </collision>
    <!-- Mid beam -->
    <collision name="mid_left_beam_lower">
      <origin rpy="0.1 0.25 0" xyz="-0.65 0.99 0.7"/>
      <geometry>
        <cylinder length="0.45" radius="0.03"/>
      </geometry>
    </collision>
    <collision name="mid_left_beam_medium">
      <origin rpy="0.75 0.25 0" xyz="-0.57 0.87 1.05"/>
      <geometry>
        <cylinder length="0.32" radius="0.03"/>
      </geometry>
    </collision>
    <collision name="mid_left_beam_upper">
      <origin rpy="1.35 0.25 0" xyz="-0.55 0.65 1.17"/>
      <geometry>
        <cylinder length="0.3" radius="0.03"/>
      </geometry>
    </collision>
    <!-- Rear beam -->
    <collision name="rear_left_beam_lower">
      <origin rpy="0 -0.15 0" xyz="-0.74 1.03 0.7"/>
      <geometry>
        <cylinder length="0.45" radius="0.03"/>
      </geometry>
    </collision>
    <collision name="rear_left_beam_medium">
      <origin rpy="0.75 -0.15 0" xyz="-0.79 0.91 1.05"/>
      <geometry>
        <cylinder length="0.32" radius="0.03"/>
      </geometry>
    </collision>
    <collision name="rear_left_beam_upper">
      <origin rpy="1.45 -0.15 0" xyz="-0.81 0.67 1.18"/>
      <geometry>
        <cylinder length="0.3" radius="0.03"/>
      </geometry>
    </collision>
    <!-- Joint -->
    <collision name="left_joint">
      <origin rpy="0 -0.6 0" xyz="0.58 1.03 0.6"/>
      <geometry>
        <box size="0.65 0.2 0.1"/>
      </geometry>
    </collision>
    <!-- Main float -->
    <collision name="right_float">
      <origin rpy="0 1.57 0" xyz="-0.4 -1.03 0.2"/>
      <geometry>
        <cylinder length="4" radius="0.2"/>
      </geometry>
    </collision>
    <collision name="right_mid_float">
      <origin rpy="0 1.38 0" xyz="1.85 -1.03 0.3"/>
      <geometry>
        <cylinder length="0.5" radius="0.17"/>
      </geometry>
    </collision>
    <collision name="right_front_float">
      <origin rpy="0 1.3 0" xyz="2.3 -1.03 0.4"/>
      <geometry>
        <cylinder length="0.45" radius="0.12"/>
      </geometry>
    </collision>
    <!-- Front beam -->
    <collision name="front_right_beam_lower">
      <origin rpy="-0.78 0 0" xyz="0.9 -0.85 1"/>
      <geometry>
        <cylinder length="0.5" radius="0.04"/>
      </geometry>
    </collision>
    <collision name="front_right_beam_upper">
      <origin rpy="-1.35 0 0" xyz="0.9 -0.6 1.18"/>
      <geometry>
        <cylinder length="0.2" radius="0.04"/>
      </geometry>
    </collision>
    <!-- Mid beam -->
    <collision name="mid_right_beam_lower">
      <origin rpy="-0.1 0.25 0" xyz="-0.65 -0.99 0.7"/>
      <geometry>
        <cylinder length="0.45" radius="0.03"/>
      </geometry>
    </collision>
    <collision name="mid_right_beam_medium">
      <origin rpy="-0.75 0.25 0" xyz="-0.57 -0.87 1.05"/>
      <geometry>
        <cylinder length="0.32" radius="0.03"/>
      </geometry>
    </collision>
    <collision name="mid_right_beam_upper">
      <origin rpy="-1.35 0.25 0" xyz="-0.55 -0.65 1.17"/>
      <geometry>
        <cylinder length="0.3" radius="0.03"/>
      </geometry>
    </collision>
    <!-- Rear beam -->
    <collision name="rear_right_beam_lower">
      <origin rpy="0 -0.15 0" xyz="-0.74 -1.03 0.7"/>
      <geometry>
        <cylinder length="0.45" radius="0.03"/>
      </geometry>
    </collision>
    <collision name="rear_right_beam_medium">
      <origin rpy="-0.75 -0.15 0" xyz="-0.79 -0.91 1.05"/>
      <geometry>
        <cylinder length="0.32" radius="0.03"/>
      </geometry>
    </collision>
    <collision name="rear_right_beam_upper">
      <origin rpy="-1.45 -0.15 0" xyz="-0.81 -0.67 1.18"/>
      <geometry>
        <cylinder length="0.3" radius="0.03"/>
      </geometry>
    </collision>
    <!-- Joint -->
    <collision name="right_joint">
      <origin rpy="0 -0.6 0" xyz="0.58 -1.03 0.6"/>
      <geometry>
        <box size="0.65 0.2 0.1"/>
      </geometry>
    </collision>
    <!-- Top base -->
    <collision name="top_base">
      <origin rpy="0 0 0" xyz="0 -0 1.25"/>
      <geometry>
        <box size="1.85 1 0.1"/>
      </geometry>
    </collision>
    <inertial>
      <!-- From WAM-V spec sheet -->
      <mass value="180.0"/>
      <inertia ixx="120.0" ixy="0.0" ixz="0.0" iyy="393" iyz="0.0" izz="446.0"/>
    </inertial>
  </link>
  <joint name="wamv/dummy_joint2" type="fixed">
    <parent link="wamv/dummy_link"/>
    <child link="wamv/qr_code"/>
  </joint>
  <link name="wamv/qr_code">
    <visual>
      <geometry>
        <box size="0.6 0.6 0.01"/>
      </geometry>
      <origin rpy="0 0 0" xyz="0 0 1.30"/>
      <material name="qr">
        <texture filename="file:///home/clp/catkin_ws/src/auto_landing/models/boat_vrx/textures/landing_pad.PNG"/>
        <!-- <script>
          <uri>model://my_landing_pad/materials/scripts</uri>
          <uri>model://my_landing_pad/materials/textures</uri>
          <name>Cube/my_landing_pad</name>
        </script> -->
      </material>
    </visual>
  </link>
  <link name="wamv/left_engine_link">
    <visual>
      <geometry>
        <mesh filename="package://wamv_description/models/engine/mesh/engine.dae"/>
      </geometry>
    </visual>
    <collision name="left_engine_vertical_axis_collision">
      <origin rpy="0 0 0" xyz="-0.16 0 -0.24"/>
      <geometry>
        <box size="0.2 0.15 0.83"/>
      </geometry>
    </collision>
    <collision name="left_engine_rear_end_collision">
      <origin rpy="0 0 0" xyz="-0.34 0 0.12"/>
      <geometry>
        <box size="0.12 0.15 0.12"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="15"/>
      <inertia ixx="0.889245" ixy="0.0" ixz="0.0" iyy="0.911125" iyz="0.0" izz="0.078125"/>
    </inertial>
  </link>
  <link name="wamv/left_propeller_link">
    <visual>
      <geometry>
        <mesh filename="package://wamv_description/models/propeller/mesh/propeller.dae"/>
      </geometry>
    </visual>
    <collision name="left_propeller_collision">
      <origin rpy="0 1.57 0" xyz="-0.08 0 0"/>
      <geometry>
        <cylinder length="0.18" radius="0.24"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="0.5"/>
      <inertia ixx="0.008545" ixy="0.0" ixz="0.0" iyy="0.008545" iyz="0.0" izz="0.0144"/>
    </inertial>
  </link>
  <joint name="wamv/left_chasis_engine_joint" type="revolute">
    <axis xyz="0 0 1"/>
    <limit effort="10" lower="-3.14159265359" upper="3.14159265359" velocity="10"/>
    <origin rpy="0 0 0" xyz="-2.373776 1.027135 0.318237"/>
    <parent link="wamv/base_link"/>
    <child link="wamv/left_engine_link"/>
  </joint>
  <joint name="wamv/left_engine_propeller_joint" type="continuous">
    <axis rpy="0 0 0" xyz="1 0 0"/>
    <parent link="wamv/left_engine_link"/>
    <child link="wamv/left_propeller_link"/>
    <origin rpy="0 0 0" xyz="-0.278156 0 -0.509371"/>
    <limit effort="100" velocity="100"/>
    <dynamics damping="0.05" friction="0.05"/>
  </joint>
  <link name="wamv/right_engine_link">
    <visual>
      <geometry>
        <mesh filename="package://wamv_description/models/engine/mesh/engine.dae"/>
      </geometry>
    </visual>
    <collision name="right_engine_vertical_axis_collision">
      <origin rpy="0 0 0" xyz="-0.16 0 -0.24"/>
      <geometry>
        <box size="0.2 0.15 0.83"/>
      </geometry>
    </collision>
    <collision name="right_engine_rear_end_collision">
      <origin rpy="0 0 0" xyz="-0.34 0 0.12"/>
      <geometry>
        <box size="0.12 0.15 0.12"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="15"/>
      <inertia ixx="0.889245" ixy="0.0" ixz="0.0" iyy="0.911125" iyz="0.0" izz="0.078125"/>
    </inertial>
  </link>
  <link name="wamv/right_propeller_link">
    <visual>
      <geometry>
        <mesh filename="package://wamv_description/models/propeller/mesh/propeller.dae"/>
      </geometry>
    </visual>
    <collision name="right_propeller_collision">
      <origin rpy="0 1.57 0" xyz="-0.08 0 0"/>
      <geometry>
        <cylinder length="0.18" radius="0.24"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="0.5"/>
      <inertia ixx="0.008545" ixy="0.0" ixz="0.0" iyy="0.008545" iyz="0.0" izz="0.0144"/>
    </inertial>
  </link>
  <joint name="wamv/right_chasis_engine_joint" type="revolute">
    <axis xyz="0 0 1"/>
    <limit effort="10" lower="-3.14159265359" upper="3.14159265359" velocity="10"/>
    <origin rpy="0 0 0" xyz="-2.373776 -1.027135 0.318237"/>
    <parent link="wamv/base_link"/>
    <child link="wamv/right_engine_link"/>
  </joint>
  <joint name="wamv/right_engine_propeller_joint" type="continuous">
    <axis rpy="0 0 0" xyz="1 0 0"/>
    <parent link="wamv/right_engine_link"/>
    <child link="wamv/right_propeller_link"/>
    <origin rpy="0 0 0" xyz="-0.278156 0 -0.509371"/>
    <limit effort="100" velocity="100"/>
    <dynamics damping="0.05" friction="0.05"/>
  </joint>
  <gazebo>
    <plugin filename="libusv_gazebo_thrust_plugin.so" name="wamv_gazebo_thrust">
      <cmdTimeout>1.0</cmdTimeout>
      <publisherRate>100.0</publisherRate>
      <robotNamespace>wamv</robotNamespace>
      <thruster>
        <!-- Required Parameters -->
        <linkName>wamv/left_propeller_link</linkName>
        <propJointName>wamv/left_engine_propeller_joint</propJointName>
        <engineJointName>wamv/left_chasis_engine_joint</engineJointName>
        <cmdTopic>thrusters/left_thrust_cmd</cmdTopic>
        <angleTopic>thrusters/left_thrust_angle</angleTopic>
        <enableAngle>True</enableAngle>
        <!-- Optional Parameters -->
        <mappingType>1</mappingType>
        <maxCmd>1.0</maxCmd>
        <maxForceFwd>250.0</maxForceFwd>
        <maxForceRev>-100.0</maxForceRev>
        <maxAngle>1.57079632679</maxAngle>
      </thruster>
      <thruster>
        <!-- Required Parameters -->
        <linkName>wamv/right_propeller_link</linkName>
        <propJointName>wamv/right_engine_propeller_joint</propJointName>
        <engineJointName>wamv/right_chasis_engine_joint</engineJointName>
        <cmdTopic>thrusters/right_thrust_cmd</cmdTopic>
        <angleTopic>thrusters/right_thrust_angle</angleTopic>
        <enableAngle>True</enableAngle>
        <!-- Optional Parameters -->
        <mappingType>1</mappingType>
        <maxCmd>1.0</maxCmd>
        <maxForceFwd>250.0</maxForceFwd>
        <maxForceRev>-100.0</maxForceRev>
        <maxAngle>1.57079632679</maxAngle>
      </thruster>
    </plugin>
  </gazebo>
  <!--Gazebo Plugin for simulating WAM-V dynamics-->
  <gazebo>
    <plugin filename="libusv_gazebo_dynamics_plugin.so" name="usv_dynamics_wamv_dynamics_plugin">
      <bodyName>wamv/base_link</bodyName>
      <!-- Must be same as the ocean model!-->
      <waterLevel>0</waterLevel>
      <waterDensity>997.8</waterDensity>
      <!-- Added mass -->
      <xDotU>0.0</xDotU>
      <yDotV>0.0</yDotV>
      <nDotR>0.0</nDotR>
      <!-- Linear and quadratic drag -->
      <xU>51.3</xU>
      <xUU>72.4</xUU>
      <yV>40.0</yV>
      <yVV>0.0</yVV>
      <zW>500.0</zW>
      <kP>50.0</kP>
      <mQ>50.0</mQ>
      <nR>400.0</nR>
      <nRR>0.0</nRR>
      <!-- General dimensions -->
      <!--<boatArea>2.2</boatArea>-->
      <hullRadius>0.213</hullRadius>
      <boatWidth>2.4</boatWidth>
      <boatLength>4.9</boatLength>
      <!-- Length discretization, AKA, "N" -->
      <length_n>2</length_n>
      <!-- Wave model -->
      <wave_model>ocean_waves</wave_model>
    </plugin>
  </gazebo>
  <!-- Monocular Camera -->
  <link name="wamv/up_camera_link">
    <visual name="up_camera_visual">
      <origin rpy="0 0 1.57079632679" xyz="-0.033 0 0"/>
      <geometry>
        <mesh filename="package://vrx_gazebo/models/mono_camera/mesh/mono_camera.dae"/>
      </geometry>
    </visual>
    <collision name="up_camera_lens_collision">
      <origin rpy="0 1.57079632679 0" xyz="0.018 0 0"/>
      <geometry>
        <cylinder length="0.042" radius="0.015"/>
      </geometry>
    </collision>
    <collision name="up_camera_box_collision">
      <origin rpy="0 0 0" xyz="-0.0205 0 0"/>
      <geometry>
        <box size="0.036 0.03 0.03"/>
      </geometry>
    </collision>
    <!-- Model inertia as box with <size>0.078 0.03 0.03</size> -->
    <inertial>
      <mass value="0.3"/>
      <inertia ixx="0.0000450" ixy="0" ixz="0" iyy="0.0001746" iyz="0" izz="0.0001746"/>
    </inertial>
  </link>
  <!-- Define a link for the optical frame.  This will match the sensor frame and
         be oriented in the camera convention. -->
  <link name="wamv/up_camera_link_optical">
    </link>
  <!-- Sensor post -->
  <link name="wamv/up_camera_post_link">
    <visual>
      <geometry>
        <cylinder length="0.1765" radius="0.0076"/>
      </geometry>
      <material name="up_camera_post_material">
        <color rgba="0.0 0.0 0.0 1.0"/>
      </material>
    </visual>
    <collision name="up_camera_post_collision">
      <geometry>
        <cylinder length="0.1765" radius="0.0076"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="0.15885"/>
      <inertia ixx="0.000414671703375" ixy="0" ixz="0" iyy="0.000414671703375" iyz="0" izz="4.587588e-06"/>
    </inertial>
  </link>
  <!-- Sensor post arm -->
  <link name="wamv/up_camera_post_arm_link">
    <visual name="up_camera_post_arm_visual">
      <origin rpy="-1.0471975512 0 -1.57079632679" xyz="-0.038 0 -0.003"/>
      <geometry>
        <mesh filename="package://vrx_gazebo/models/sensor_post/mesh/sensor_post_arm.dae"/>
      </geometry>
    </visual>
    <collision name="up_camera_post_arm_collision">
      <origin rpy="-1.20427718388 0 -1.57079632679" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.065" radius="0.008"/>
      </geometry>
    </collision>
    <inertial>
      <origin rpy="-1.20427718388 0 -1.57079632679" xyz="0 0 0"/>
      <mass value="0.1"/>
      <inertia ixx="0.00003680833" ixy="0" ixz="0" iyy="0.00003680833" iyz="0" izz="0.00000320000"/>
    </inertial>
  </link>
  <joint name="wamv/base_to_up_camera_post_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0.7 0.1 1.38475"/>
    <parent link="wamv/base_link"/>
    <child link="wamv/up_camera_post_link"/>
  </joint>
  <!-- Sensor post to sensor post arm joint -->
  <joint name="wamv/up_camera_post_to_up_camera_post_arm_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0.03 0 0.08825"/>
    <parent link="wamv/up_camera_post_link"/>
    <child link="wamv/up_camera_post_arm_link"/>
  </joint>
  <!-- Sensor post arm to camera joint -->
  <joint name="wamv/up_camera_post_arm_to_up_camera_joint" type="fixed">
    <origin rpy="0 -1.57079632679 0" xyz="0.02 0 0.027"/>
    <parent link="wamv/up_camera_post_arm_link"/>
    <child link="wamv/up_camera_link"/>
  </joint>
  <!-- Camera to optical frame joint.  This is oriented to convert between ENU
         and camera conventions -->
  <joint name="wamv/up_camera_to_up_camera_link_optical_joint" type="fixed">
    <origin rpy="-1.57079632679 0 -1.57079632679" xyz="0 0 0"/>
    <parent link="wamv/up_camera_link"/>
    <child link="wamv/up_camera_link_optical"/>
  </joint>
  <!-- Gazebo camera setup -->
  <gazebo reference="wamv/up_camera_link">
    <sensor name="up_camera_sensor" type="camera">
      <update_rate>30.0</update_rate>
      <camera name="up_camera_camera">
        <horizontal_fov>1.3962634</horizontal_fov>
        <image>
          <width>1280</width>
          <height>720</height>
          <format>R8G8B8</format>
        </image>
        <clip>
          <near>0.05</near>
          <far>300</far>
        </clip>
        <noise>
          <type>gaussian</type>
          <!-- Noise is sampled independently per pixel on each frame.
                 That pixel's noise value is added to each of its color
                 channels, which at that point lie in the range [0,1]. -->
          <mean>0.0</mean>
          <stddev>0.007</stddev>
        </noise>
      </camera>
      <plugin filename="libgazebo_ros_camera.so" name="camera_plugin_up_camera">
        <alwaysOn>true</alwaysOn>
        <updateRate>0.0</updateRate>
        <cameraName>wamv/sensors/cameras/up_camera</cameraName>
        <imageTopicName>image_raw</imageTopicName>
        <cameraInfoTopicName>camera_info</cameraInfoTopicName>
        <frameName>wamv/up_camera_link_optical</frameName>
        <hackBaseline>0.07</hackBaseline>
        <distortionK1>0.0</distortionK1>
        <distortionK2>0.0</distortionK2>
        <distortionK3>0.0</distortionK3>
        <distortionT1>0.0</distortionT1>
        <distortionT2>0.0</distortionT2>
      </plugin>
    </sensor>
  </gazebo>
  <!-- Set color of post -->
  <gazebo reference="wamv/up_camera_post_link">
    <material>Gazebo/Black</material>
  </gazebo>
</robot>

